ICE: `adding a def'n for node-id NodeId(17) and def kind AnonConst but a previous def'n exists`

<!--
[31mICE[0m: Rustc ./a.rs '' 'thread 'rustc' panicked at compiler/rustc_ast_lowering/src/lib.rs:503:9: 'adding a def'n for node-id NodeId(79) and def kind AnonConst but a previous def'n exists: DefKey { parent: Some(DefIndex(10)), disambiguated_data: DisambiguatedDefPathData { data: AnonConst, disambiguator: 3 } }'', 'thread 'rustc' panicked at compiler/rustc_ast_lowering/src/lib.rs:503:9: 'adding a def'n for node-id NodeId(79) and def kind AnonConst but a previous def'n exists: DefKey { parent: Some(DefIndex(10)), disambiguated_data: DisambiguatedDefPathData { data: AnonConst, disambiguator: 3 } }''
File: /tmp/im/a.rs
-->

auto-reduced (treereduce-rust):
````rust
macro_rules! make_1 {
    () => {
        unused_braces
    };
}

fn make_1() {
    // ok
    let _: A<{ make_1!() }>; // ok
}


````

original:
````rust
//@ check-pass
//@ run-rustfix
#![warn(unused_braces)]

macro_rules! make_1 {
    () => (unused_braces)
}

struct A<const N: usize>;

fn make_1() {
    let _: A<7> // ok
    let _: A<{ 7 }>; //~ WARN unnecessary braces
    let _: A<{
    let _: A<7>; // ok
    let _: A<{ 7 }>; //~ WARN unnecessary braces
    let _: A<{ 3 + 5 }>; // ok
    let _: A<{make_1!()}>; // ok
}>; // ok
    let _: A<{make_1!()}>; // ok
}

````

Version information
````
rustc 1.83.0-nightly (0307e401c 2024-09-13)
binary: rustc
commit-hash: 0307e401c26699656ae08e3809e7d272f5c103f4
commit-date: 2024-09-13
host: x86_64-unknown-linux-gnu
release: 1.83.0-nightly
LLVM version: 19.1.0

````

Command:
`/home/matthias/.rustup/toolchains/master/bin/rustc `

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary><strong>Program output</strong></summary>
<p>

```
error[E0412]: cannot find type `A` in this scope
 --> /tmp/icemaker_global_tempdir.20nnTggI11ue/rustc_testrunner_tmpdir_reporting.YvPvHHqLRYEe/mvce.rs:9:12
  |
9 |     let _: A<{ make_1!() }>; // ok
  |            ^ not found in this scope

error[E0425]: cannot find value `unused_braces` in this scope
 --> /tmp/icemaker_global_tempdir.20nnTggI11ue/rustc_testrunner_tmpdir_reporting.YvPvHHqLRYEe/mvce.rs:3:9
  |
3 |         unused_braces
  |         ^^^^^^^^^^^^^ not found in this scope
...
9 |     let _: A<{ make_1!() }>; // ok
  |                --------- in this macro invocation
  |
  = note: this error originates in the macro `make_1` (in Nightly builds, run with -Z macro-backtrace for more info)
help: you might be missing a const parameter
  |
7 | fn make_1<const unused_braces: /* Type */>() {
  |          +++++++++++++++++++++++++++++++++

thread 'rustc' panicked at compiler/rustc_ast_lowering/src/lib.rs:503:9:
adding a def'n for node-id NodeId(17) and def kind AnonConst but a previous def'n exists: DefKey { parent: Some(DefIndex(4)), disambiguated_data: DisambiguatedDefPathData { data: AnonConst, disambiguator: 0 } }
stack backtrace:
   0:     0x792523c73fba - <std::sys::backtrace::BacktraceLock::print::DisplayBacktrace as core::fmt::Display>::fmt::hfb6d8fb4f154a1fc
   1:     0x792524403257 - core::fmt::write::h7808e5e403c5f075
   2:     0x792525357af3 - std::io::Write::write_fmt::h538db9e40c2a2888
   3:     0x792523c73e12 - std::sys::backtrace::BacktraceLock::print::he7092e8746cba8fb
   4:     0x792523c76591 - std::panicking::default_hook::{{closure}}::hb4d26de5c9fcbcdc
   5:     0x792523c763c4 - std::panicking::default_hook::h810d992f1f481ea3
   6:     0x792522d7715f - std[b0815b464da53495]::panicking::update_hook::<alloc[2565d1873fda1868]::boxed::Box<rustc_driver_impl[8162622be99fcd85]::install_ice_hook::{closure#0}>>::{closure#0}
   7:     0x792523c76cb8 - std::panicking::rust_panic_with_hook::h678fb9ba687f252d
   8:     0x792523c76a87 - std::panicking::begin_panic_handler::{{closure}}::hdf1255a7653694ee
   9:     0x792523c74469 - std::sys::backtrace::__rust_end_short_backtrace::h559381c391964077
  10:     0x792523c76754 - rust_begin_unwind
  11:     0x792520a94743 - core::panicking::panic_fmt::h3dc6c16be86388a1
  12:     0x792525db6c35 - <rustc_ast_lowering[8c3fea5c94799a38]::LoweringContext>::create_def.cold
  13:     0x792524ac59a6 - <rustc_ast_lowering[8c3fea5c94799a38]::LoweringContext>::lower_anon_const_to_anon_const
  14:     0x792524ac563e - <rustc_ast_lowering[8c3fea5c94799a38]::LoweringContext>::lower_anon_const_to_const_arg_direct
  15:     0x792524abb7ee - <rustc_ast_lowering[8c3fea5c94799a38]::LoweringContext>::lower_qpath
  16:     0x792524ac7937 - <rustc_ast_lowering[8c3fea5c94799a38]::LoweringContext>::lower_ty_direct
  17:     0x792524ad2fc2 - <rustc_ast_lowering[8c3fea5c94799a38]::LoweringContext>::lower_block_noalloc
  18:     0x792524a99eec - <rustc_ast_lowering[8c3fea5c94799a38]::LoweringContext>::lower_maybe_coroutine_body
  19:     0x792524a8ac06 - <rustc_ast_lowering[8c3fea5c94799a38]::LoweringContext>::lower_item_kind
  20:     0x792524a9c833 - <rustc_ast_lowering[8c3fea5c94799a38]::item::ItemLowerer>::lower_node
  21:     0x792524a9f57f - rustc_ast_lowering[8c3fea5c94799a38]::lower_to_hir
  22:     0x792525308c20 - rustc_query_impl[83f8432e820c675c]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[83f8432e820c675c]::query_impl::hir_crate::dynamic_query::{closure#2}::{closure#0}, rustc_middle[759127956bc2125]::query::erase::Erased<[u8; 8usize]>>
  23:     0x792525307b39 - rustc_query_system[6df6dc7cefe5b4ac]::query::plumbing::try_execute_query::<rustc_query_impl[83f8432e820c675c]::DynamicConfig<rustc_query_system[6df6dc7cefe5b4ac]::query::caches::SingleCache<rustc_middle[759127956bc2125]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[83f8432e820c675c]::plumbing::QueryCtxt, false>
  24:     0x79252530775e - rustc_query_impl[83f8432e820c675c]::query_impl::hir_crate::get_query_non_incr::__rust_end_short_backtrace
  25:     0x792524616985 - rustc_query_impl[83f8432e820c675c]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[83f8432e820c675c]::query_impl::hir_attrs::dynamic_query::{closure#2}::{closure#0}, rustc_middle[759127956bc2125]::query::erase::Erased<[u8; 8usize]>>
  26:     0x792524615f22 - rustc_query_system[6df6dc7cefe5b4ac]::query::plumbing::try_execute_query::<rustc_query_impl[83f8432e820c675c]::DynamicConfig<rustc_query_system[6df6dc7cefe5b4ac]::query::caches::VecCache<rustc_hir[26e18d3688e9fba1]::hir_id::OwnerId, rustc_middle[759127956bc2125]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[83f8432e820c675c]::plumbing::QueryCtxt, false>
  27:     0x792524615c8f - rustc_query_impl[83f8432e820c675c]::query_impl::hir_attrs::get_query_non_incr::__rust_end_short_backtrace
  28:     0x7925246c3502 - <rustc_middle[759127956bc2125]::hir::map::Map>::attrs
  29:     0x79252531cf01 - rustc_passes[507916cbcf1ee5a8]::entry::entry_fn
  30:     0x79252531ce98 - rustc_query_impl[83f8432e820c675c]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[83f8432e820c675c]::query_impl::entry_fn::dynamic_query::{closure#2}::{closure#0}, rustc_middle[759127956bc2125]::query::erase::Erased<[u8; 12usize]>>
  31:     0x79252531ce75 - <rustc_query_impl[83f8432e820c675c]::query_impl::entry_fn::dynamic_query::{closure#2} as core[ead5fa394faa3a31]::ops::function::FnOnce<(rustc_middle[759127956bc2125]::ty::context::TyCtxt, ())>>::call_once
  32:     0x79252531c809 - rustc_query_system[6df6dc7cefe5b4ac]::query::plumbing::try_execute_query::<rustc_query_impl[83f8432e820c675c]::DynamicConfig<rustc_query_system[6df6dc7cefe5b4ac]::query::caches::SingleCache<rustc_middle[759127956bc2125]::query::erase::Erased<[u8; 12usize]>>, false, false, false>, rustc_query_impl[83f8432e820c675c]::plumbing::QueryCtxt, false>
  33:     0x79252531c5a9 - rustc_query_impl[83f8432e820c675c]::query_impl::entry_fn::get_query_non_incr::__rust_end_short_backtrace
  34:     0x792524ca4983 - rustc_interface[c4b0133a9ab33d3c]::passes::run_required_analyses
  35:     0x79252518a09e - rustc_interface[c4b0133a9ab33d3c]::passes::analysis
  36:     0x79252518a071 - rustc_query_impl[83f8432e820c675c]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[83f8432e820c675c]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[759127956bc2125]::query::erase::Erased<[u8; 1usize]>>
  37:     0x79252530556e - rustc_query_system[6df6dc7cefe5b4ac]::query::plumbing::try_execute_query::<rustc_query_impl[83f8432e820c675c]::DynamicConfig<rustc_query_system[6df6dc7cefe5b4ac]::query::caches::SingleCache<rustc_middle[759127956bc2125]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[83f8432e820c675c]::plumbing::QueryCtxt, false>
  38:     0x7925253052cf - rustc_query_impl[83f8432e820c675c]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  39:     0x79252517bffc - rustc_interface[c4b0133a9ab33d3c]::interface::run_compiler::<core[ead5fa394faa3a31]::result::Result<(), rustc_span[6d3b9e60beb0a608]::ErrorGuaranteed>, rustc_driver_impl[8162622be99fcd85]::run_compiler::{closure#0}>::{closure#1}
  40:     0x79252526a710 - std[b0815b464da53495]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[c4b0133a9ab33d3c]::util::run_in_thread_with_globals<rustc_interface[c4b0133a9ab33d3c]::util::run_in_thread_pool_with_globals<rustc_interface[c4b0133a9ab33d3c]::interface::run_compiler<core[ead5fa394faa3a31]::result::Result<(), rustc_span[6d3b9e60beb0a608]::ErrorGuaranteed>, rustc_driver_impl[8162622be99fcd85]::run_compiler::{closure#0}>::{closure#1}, core[ead5fa394faa3a31]::result::Result<(), rustc_span[6d3b9e60beb0a608]::ErrorGuaranteed>>::{closure#0}, core[ead5fa394faa3a31]::result::Result<(), rustc_span[6d3b9e60beb0a608]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[ead5fa394faa3a31]::result::Result<(), rustc_span[6d3b9e60beb0a608]::ErrorGuaranteed>>
  41:     0x79252526ad7a - <<std[b0815b464da53495]::thread::Builder>::spawn_unchecked_<rustc_interface[c4b0133a9ab33d3c]::util::run_in_thread_with_globals<rustc_interface[c4b0133a9ab33d3c]::util::run_in_thread_pool_with_globals<rustc_interface[c4b0133a9ab33d3c]::interface::run_compiler<core[ead5fa394faa3a31]::result::Result<(), rustc_span[6d3b9e60beb0a608]::ErrorGuaranteed>, rustc_driver_impl[8162622be99fcd85]::run_compiler::{closure#0}>::{closure#1}, core[ead5fa394faa3a31]::result::Result<(), rustc_span[6d3b9e60beb0a608]::ErrorGuaranteed>>::{closure#0}, core[ead5fa394faa3a31]::result::Result<(), rustc_span[6d3b9e60beb0a608]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[ead5fa394faa3a31]::result::Result<(), rustc_span[6d3b9e60beb0a608]::ErrorGuaranteed>>::{closure#1} as core[ead5fa394faa3a31]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  42:     0x79252526b16b - std::sys::pal::unix::thread::Thread::new::thread_start::h5b72917264317154
  43:     0x79252694f39d - <unknown>
  44:     0x7925269d449c - <unknown>
  45:                0x0 - <unknown>

error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: rustc 1.83.0-nightly (0307e401c 2024-09-13) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [hir_crate] getting the crate HIR
#1 [hir_attrs] getting HIR owner attributes in ``
end of query stack
error: aborting due to 2 previous errors

Some errors have detailed explanations: E0412, E0425.
For more information about an error, try `rustc --explain E0412`.

```

</p>
</details>

<!--
query stack:
#0 [hir_crate] getting the crate HIR
#1 [hir_attrs] getting HIR owner attributes in ``

-->

--- a/compiler/rustc_ast/src/ast.rs
+++ b/compiler/rustc_ast/src/ast.rs
@@ -1187,8 +1187,8 @@ impl Expr {
     /// `min_const_generics` as more complex expressions are not supported.
     ///
     /// Does not ensure that the path resolves to a const param, the caller should check this.
-    pub fn is_potential_trivial_const_arg(&self) -> bool {
-        let this = self.maybe_unwrap_block();
+    pub fn is_potential_trivial_const_arg(&self, strip_identity_block: bool) -> bool {
+        let this = if strip_identity_block { self.maybe_unwrap_block().1 } else { self };
 
         if let ExprKind::Path(None, path) = &this.kind
             && path.is_potential_trivial_const_arg()
@@ -1199,14 +1199,15 @@ impl Expr {
         }
     }
 
-    pub fn maybe_unwrap_block(&self) -> &Expr {
+    /// Returns an expression with (when possible) *one* outter brace removed
+    pub fn maybe_unwrap_block(&self) -> (bool, &Expr) {
         if let ExprKind::Block(block, None) = &self.kind
             && let [stmt] = block.stmts.as_slice()
             && let StmtKind::Expr(expr) = &stmt.kind
         {
-            expr
+            (true, expr)
         } else {
-            self
+            (false, self)
         }
     }
 

--- a/compiler/rustc_ast_lowering/src/asm.rs
+++ b/compiler/rustc_ast_lowering/src/asm.rs
@@ -220,7 +220,7 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
                             let parent_def_id = self.current_def_id_parent;
                             let node_id = self.next_node_id();
                             // HACK(min_generic_const_args): see lower_anon_const
-                            if !expr.is_potential_trivial_const_arg() {
+                            if !expr.is_potential_trivial_const_arg(true) {
                                 self.create_def(
                                     parent_def_id,
                                     node_id,

--- a/compiler/rustc_ast_lowering/src/expr.rs
+++ b/compiler/rustc_ast_lowering/src/expr.rs
@@ -387,7 +387,7 @@ impl<'hir> LoweringContext<'_, 'hir> {
                 let node_id = self.next_node_id();
 
                 // HACK(min_generic_const_args): see lower_anon_const
-                if !arg.is_potential_trivial_const_arg() {
+                if !arg.is_potential_trivial_const_arg(true) {
                     // Add a definition for the in-band const def.
                     self.create_def(parent_def_id, node_id, kw::Empty, DefKind::AnonConst, f.span);
                 }

--- a/compiler/rustc_ast_lowering/src/lib.rs
+++ b/compiler/rustc_ast_lowering/src/lib.rs
@@ -2441,7 +2441,7 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
     /// See [`hir::ConstArg`] for when to use this function vs
     /// [`Self::lower_anon_const_to_const_arg`].
     fn lower_anon_const_to_anon_const(&mut self, c: &AnonConst) -> &'hir hir::AnonConst {
-        if c.value.is_potential_trivial_const_arg() {
+        if c.value.is_potential_trivial_const_arg(true) {
             // HACK(min_generic_const_args): see DefCollector::visit_anon_const
             // Over there, we guess if this is a bare param and only create a def if
             // we think it's not. However we may can guess wrong (see there for example)

--- a/compiler/rustc_resolve/src/def_collector.rs
+++ b/compiler/rustc_resolve/src/def_collector.rs
@@ -138,6 +138,61 @@ impl<'a, 'ra, 'tcx> DefCollector<'a, 'ra, 'tcx> {
         );
         assert!(old_parent.is_none(), "parent `LocalDefId` is reset for an invocation");
     }
+
+    /// Determines whether the const argument `AnonConst` is a simple macro call, optionally
+    /// surrounded with braces.
+    ///
+    /// If this const argument *is* a trivial macro call then the id for the macro call is
+    /// returned along with the information required to build the anon const's def if
+    /// the macro call expands to a non-trivial expression.
+    fn is_const_arg_trivial_macro_expansion(
+        &self,
+        anon_const: &'a AnonConst,
+    ) -> Option<(PendingAnonConstInfo, NodeId)> {
+        let (block_was_stripped, expr) = anon_const.value.maybe_unwrap_block();
+        match expr {
+            Expr { kind: ExprKind::MacCall(..), id, .. } => Some((
+                PendingAnonConstInfo {
+                    id: anon_const.id,
+                    span: anon_const.value.span,
+                    block_was_stripped,
+                },
+                *id,
+            )),
+            _ => None,
+        }
+    }
+
+    /// Determines whether the expression `const_arg_sub_expr` is a simple macro call, sometimes
+    /// surrounded with braces if a set of braces has not already been entered. This is required
+    /// as `{ N }` is treated as equivalent to a bare parameter `N` whereas `{{ N }}` is treated as
+    /// a real block expression and is lowered to an anonymous constant which is not allowed to use
+    /// generic parameters.
+    ///
+    /// If this expression is a trivial macro call then the id for the macro call is
+    /// returned along with the information required to build the anon const's def if
+    /// the macro call expands to a non-trivial expression.
+    fn is_const_arg_sub_expr_trivial_macro_expansion(
+        &self,
+        const_arg_sub_expr: &'a Expr,
+    ) -> Option<(PendingAnonConstInfo, NodeId)> {
+        let pending_anon = self.pending_anon_const_info.unwrap_or_else(||
+            panic!("Checking expr is trivial macro call without having entered anon const: `{const_arg_sub_expr:?}`"),
+        );
+
+        let (block_was_stripped, expr) = if pending_anon.block_was_stripped {
+            (true, const_arg_sub_expr)
+        } else {
+            const_arg_sub_expr.maybe_unwrap_block()
+        };
+
+        match expr {
+            Expr { kind: ExprKind::MacCall(..), id, .. } => {
+                Some((PendingAnonConstInfo { block_was_stripped, ..pending_anon }, *id))
+            }
+            _ => None,
+        }
+    }
 }
 
 impl<'a, 'ra, 'tcx> visit::Visitor<'a> for DefCollector<'a, 'ra, 'tcx> {
@@ -354,12 +409,12 @@ impl<'a, 'ra, 'tcx> visit::Visitor<'a> for DefCollector<'a, 'ra, 'tcx> {
         // items will be messed up, but that's ok because there can't be any if we're just looking
         // for bare idents.
 
-        if matches!(constant.value.maybe_unwrap_block().kind, ExprKind::MacCall(..)) {
-            // See self.pending_anon_const_info for explanation
-            self.pending_anon_const_info =
-                Some(PendingAnonConstInfo { id: constant.id, span: constant.value.span });
-            return visit::walk_anon_const(self, constant);
-        } else if constant.value.is_potential_trivial_const_arg() {
+        if let Some((pending_anon, macro_invoc)) =
+            self.is_const_arg_trivial_macro_expansion(constant)
+        {
+            self.pending_anon_const_info = Some(pending_anon);
+            return self.visit_macro_invoc(macro_invoc);
+        } else if constant.value.is_potential_trivial_const_arg(true) {
             return visit::walk_anon_const(self, constant);
         }
 
@@ -368,23 +423,36 @@ impl<'a, 'ra, 'tcx> visit::Visitor<'a> for DefCollector<'a, 'ra, 'tcx> {
     }
 
     fn visit_expr(&mut self, expr: &'a Expr) {
-        if matches!(expr.kind, ExprKind::MacCall(..)) {
-            return self.visit_macro_invoc(expr.id);
+        // If we're visiting the expression of a const argument that was a macro call then
+        // check if it is *still* unknown whether it is a trivial const arg or not. If so
+        // recurse into the macro call and delay creating the anon const def until expansion.
+        if self.pending_anon_const_info.is_some()
+            && let Some((pending_anon, macro_invoc)) =
+                self.is_const_arg_sub_expr_trivial_macro_expansion(expr)
+        {
+            self.pending_anon_const_info = Some(pending_anon);
+            return self.visit_macro_invoc(macro_invoc);
         }
 
-        let grandparent_def = if let Some(pending_anon) = self.pending_anon_const_info.take() {
-            // See self.pending_anon_const_info for explanation
-            if !expr.is_potential_trivial_const_arg() {
+        // See self.pending_anon_const_info for explanation
+        let parent_def = self
+            .pending_anon_const_info
+            .take()
+            // If we already stripped away a set of braces then do not do it again when determining
+            // if the macro expanded to a trivial const arg. This arises in cases such as:
+            // `Foo<{ bar!() }>` where `bar!()` expands to `{ N }`. This should not be considered a
+            // trivial const argument even though `{ N }` by itself *is*.
+            .filter(|pending_anon| {
+                !expr.is_potential_trivial_const_arg(!pending_anon.block_was_stripped)
+            })
+            .map(|pending_anon| {
                 self.create_def(pending_anon.id, kw::Empty, DefKind::AnonConst, pending_anon.span)
-            } else {
-                self.parent_def
-            }
-        } else {
-            self.parent_def
-        };
+            })
+            .unwrap_or(self.parent_def);
 
-        self.with_parent(grandparent_def, |this| {
+        self.with_parent(parent_def, |this| {
             let parent_def = match expr.kind {
+                ExprKind::MacCall(..) => return this.visit_macro_invoc(expr.id),
                 ExprKind::Closure(..) | ExprKind::Gen(..) => {
                     this.create_def(expr.id, kw::Empty, DefKind::Closure, expr.span)
                 }

--- a/compiler/rustc_resolve/src/late.rs
+++ b/compiler/rustc_resolve/src/late.rs
@@ -4524,7 +4524,7 @@ impl<'a, 'ast, 'ra: 'ast, 'tcx> LateResolutionVisitor<'a, 'ast, 'ra, 'tcx> {
         );
 
         self.resolve_anon_const_manual(
-            constant.value.is_potential_trivial_const_arg(),
+            constant.value.is_potential_trivial_const_arg(true),
             anon_const_kind,
             |this| this.resolve_expr(&constant.value, None),
         )
@@ -4688,7 +4688,7 @@ impl<'a, 'ast, 'ra: 'ast, 'tcx> LateResolutionVisitor<'a, 'ast, 'ra, 'tcx> {
                     // that is how they will be later lowered to HIR.
                     if const_args.contains(&idx) {
                         self.resolve_anon_const_manual(
-                            argument.is_potential_trivial_const_arg(),
+                            argument.is_potential_trivial_const_arg(true),
                             AnonConstKind::ConstArg(IsRepeatExpr::No),
                             |this| this.resolve_expr(argument, None),
                         );

--- a/compiler/rustc_resolve/src/lib.rs
+++ b/compiler/rustc_resolve/src/lib.rs
@@ -190,6 +190,11 @@ impl InvocationParent {
 
 #[derive(Copy, Debug, Clone)]
 struct PendingAnonConstInfo {
+    // A const arg is only a "trivial" const arg if it has at *most* one set of braces
+    // around the argument. We track whether we have stripped an outter brace so that
+    // if a macro expands to a braced expression *and* the macro was itself inside of
+    // some braces then we can consider it to be a non-trivial const argument.
+    block_was_stripped: bool,
     id: NodeId,
     span: Span,
 }

--- a/tests/ui/const-generics/early/trivial-const-arg-macro-braced-expansion.rs
+++ b/tests/ui/const-generics/early/trivial-const-arg-macro-braced-expansion.rs
@@ -1,0 +1,14 @@
+macro_rules! y {
+    () => {
+        N
+    };
+}
+
+struct A<const N: usize>;
+
+fn foo<const N: usize>() -> A<{ y!() }> {
+    A::<1>
+    //~^ ERROR: mismatched types
+}
+
+fn main() {}

--- a/tests/ui/const-generics/early/trivial-const-arg-macro-braced-expansion.stderr
+++ b/tests/ui/const-generics/early/trivial-const-arg-macro-braced-expansion.stderr
@@ -1,0 +1,14 @@
+error[E0308]: mismatched types
+  --> $DIR/trivial-const-arg-macro-braced-expansion.rs:10:5
+   |
+LL | fn foo<const N: usize>() -> A<{ y!() }> {
+   |                             ----------- expected `A<N>` because of return type
+LL |     A::<1>
+   |     ^^^^^^ expected `N`, found `1`
+   |
+   = note: expected struct `A<N>`
+              found struct `A<1>`
+
+error: aborting due to 1 previous error
+
+For more information about this error, try `rustc --explain E0308`.

--- a/tests/ui/const-generics/early/trivial-const-arg-macro-nested-braces-2.rs
+++ b/tests/ui/const-generics/early/trivial-const-arg-macro-nested-braces-2.rs
@@ -1,0 +1,15 @@
+macro_rules! y {
+    () => {
+        N
+        //~^ ERROR: generic parameters may not be used in const operations
+    };
+}
+
+struct A<const N: usize>;
+
+#[rustfmt::skip]
+fn foo<const N: usize>() -> A<{{ y!() }}> {
+    A::<1>
+}
+
+fn main() {}

--- a/tests/ui/const-generics/early/trivial-const-arg-macro-nested-braces-2.stderr
+++ b/tests/ui/const-generics/early/trivial-const-arg-macro-nested-braces-2.stderr
@@ -1,0 +1,15 @@
+error: generic parameters may not be used in const operations
+  --> $DIR/trivial-const-arg-macro-nested-braces-2.rs:3:9
+   |
+LL |         N
+   |         ^ cannot perform const operation using `N`
+...
+LL | fn foo<const N: usize>() -> A<{{ y!() }}> {
+   |                                  ---- in this macro invocation
+   |
+   = help: const parameters may only be used as standalone arguments, i.e. `N`
+   = help: add `#![feature(generic_const_exprs)]` to allow generic const expressions
+   = note: this error originates in the macro `y` (in Nightly builds, run with -Z macro-backtrace for more info)
+
+error: aborting due to 1 previous error
+

--- a/tests/ui/const-generics/early/trivial-const-arg-macro-nested-braces.rs
+++ b/tests/ui/const-generics/early/trivial-const-arg-macro-nested-braces.rs
@@ -1,0 +1,15 @@
+#[rustfmt::skip]
+macro_rules! y {
+    () => {
+        { N }
+        //~^ ERROR: generic parameters may not be used in const operations
+    };
+}
+
+struct A<const N: usize>;
+
+fn foo<const N: usize>() -> A<{ y!() }> {
+    A::<1>
+}
+
+fn main() {}

--- a/tests/ui/const-generics/early/trivial-const-arg-macro-nested-braces.stderr
+++ b/tests/ui/const-generics/early/trivial-const-arg-macro-nested-braces.stderr
@@ -1,0 +1,15 @@
+error: generic parameters may not be used in const operations
+  --> $DIR/trivial-const-arg-macro-nested-braces.rs:4:11
+   |
+LL |         { N }
+   |           ^ cannot perform const operation using `N`
+...
+LL | fn foo<const N: usize>() -> A<{ y!() }> {
+   |                                 ---- in this macro invocation
+   |
+   = help: const parameters may only be used as standalone arguments, i.e. `N`
+   = help: add `#![feature(generic_const_exprs)]` to allow generic const expressions
+   = note: this error originates in the macro `y` (in Nightly builds, run with -Z macro-backtrace for more info)
+
+error: aborting due to 1 previous error
+

--- a/tests/ui/const-generics/early/trivial-const-arg-nested-braces.rs
+++ b/tests/ui/const-generics/early/trivial-const-arg-nested-braces.rs
@@ -1,0 +1,9 @@
+struct A<const N: usize>;
+
+#[rustfmt::skip]
+fn foo<const N: usize>() -> A<{ { N } }> {
+    //~^ ERROR: generic parameters may not be used in const operations
+    A::<1>
+}
+
+fn main() {}

--- a/tests/ui/const-generics/early/trivial-const-arg-nested-braces.stderr
+++ b/tests/ui/const-generics/early/trivial-const-arg-nested-braces.stderr
@@ -1,0 +1,11 @@
+error: generic parameters may not be used in const operations
+  --> $DIR/trivial-const-arg-nested-braces.rs:4:35
+   |
+LL | fn foo<const N: usize>() -> A<{ { N } }> {
+   |                                   ^ cannot perform const operation using `N`
+   |
+   = help: const parameters may only be used as standalone arguments, i.e. `N`
+   = help: add `#![feature(generic_const_exprs)]` to allow generic const expressions
+
+error: aborting due to 1 previous error
+

